### ISSUE TITLE
Allow probes to run on a more granular timer.

### ISSUE THREAD

**Issue by markusthoemmes** [kubernetes/kubernetes#76951]
**What would you like to be added**:

For posterity:

```golang
type Probe struct {
	InitialDelaySeconds int32
	TimeoutSeconds int32
	PeriodSeconds int32
	...
}
```

Probes today take all their specified timeouts and delays in **seconds**. I feel like that arbitrarily sets a lower bound on pod readiness of at least 1 second. For usual operations of K8s workloads, that's all fine. When looking at serverless workloads, where a quick startup of new containers is key and directly reflected in the end user's latency.

Can these be `Duration` instead, so the user can define the granularity like she pleases?

**Why is this needed**:

To make workloads with very sensitive startup latency (like serverless workloads) easier to implement on Kubernetes.

---

Sorry if this is a duplicate, I haven't found a similar issue.

**Comment by evankanderson** [comment]
@kubernetes/sig-node-feature-requests

**Comment by liggitt** [comment]
If a more granular duration field was added, the definition of the field should include the units, not tie a field to golang duration parsing. c.f. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md

**Comment by dgerd** [comment]
/remove-lifecycle stale

@liggitt We have worked around this in Knative by rewriting probes and having a sidecar do the probing. It would be great to at some point in time in the future undo this and go back to using K8s probes. What needs to be done on our side to drive this conversation forward?

**Comment by evankanderson** [comment]
Knative has ended up reimplementing readiness probes and Endpoint updates to enable sub-second Pod readiness during scale-up events.

@vagababov for visibility on the upstream issue

(this seems to be a dup of closed issue #38778, which didn't seem to have a specific need and aged out)

**Comment by evankanderson** [comment]
@maximilien was interested in this today.

**Comment by MHBauer** [comment]
https://github.com/kubernetes/kubernetes/issues/76951#issuecomment-486465014

Trying to think through this.
With the existing type declarations:

1. Can we compatibly swap out int32 for a resource.Quantity and some code? Or does it require a v2 of the Probe struct?

2. Do we add new optional fields that describe some suitable subdivision of a second (millis, nanos)? Then change some defaulting if the subdivision is set without setting the larger whole seconds. Logic TBD.
 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go

I fear this might interact with some of the ongoing sidecar work, but not sure yet 😅.

 I can write it up a bit more and raise it in sig-node before we go full KEP on it.

**Comment by liggitt** [comment]
> 1. Can we compatibly swap out int32 for a resource.Quantity and some code?

no, string values are not backward-compatible with clients expecting int values

> 2. Do we add new optional fields that describe some suitable subdivision of a second (millis, nanos)? Then change some defaulting if the subdivision is set without setting the larger whole seconds

An approach that allows 0 into a field that previously defaulted to 1 and validated a minimum value of 1 will be difficult to roll out compatibly. For example:
* an older kubelet that doesn't know about a new fractional field could observe a value of 0 in that field and hotloop
* an n-1 API server that doesn't know about the new fractional field would default the old field to 1 on create or update, or  would treat an existing value of 0 as invalid, potentially preventing updates to the pod.

> I can write it up a bit more and raise it in sig-node before we go full KEP on it.

Are sub-second probes something we want to let API users cause the kubelet to do? That seems worth verifying with sig-node.

**Comment by derekwaynecarr** [comment]
I appreciate the motivation, but feel there are a number of other areas of potential optimization we could perform to improve pod startup latency in the kubelet->{runtime, network, storage} subsystems that allowing for shorter probe periods may be premature and introduce other unintended vectors of abuse.

**Comment by MHBauer** [comment]
Thank you both!

https://github.com/kubernetes/kubernetes/issues/76951#issuecomment-694468776
Exactly as I expected for 1 & 2. 

https://github.com/kubernetes/kubernetes/issues/76951#issuecomment-694473193
Is there a place to rally around these improvements? I will search for concrete and explicit test cases. This does appear very specific, but maybe it can drive general improvement.

**Comment by dprotaso** [comment]
/lifecycle frozen 

I believe @mikebrow is working on this

**Comment by cynepco3hahue** [comment]
@mikebrow Can I assign you if you are really working on it?

**Comment by mikebrow** [comment]
> @mikebrow Can I assign you if you are really working on it?

yes

**Comment by mikebrow** [comment]
https://github.com/kubernetes/enhancements/issues/3066
